python/cuml/metrics/pairwise_distances.pyx

```diff
@@ -279,3 +314,186 @@ def pairwise_distances(X, Y=None, metric="euclidean", handle=None,
     del Y_m
 
     return dest_m
+
+
+@cuml.internals.api_return_array(get_output_type=True)
+def sparse_pairwise_distances(X, Y=None, metric="euclidean", handle=None,
+                              convert_dtype=True, metric_arg=2, **kwds):
+    """
+    Compute the distance matrix from a vector array `X` and optional `Y`.
+
+    This method takes either one or two sparse vector arrays, and returns a
+    dense distance matrix.
+
+    If `Y` is given (default is `None`), then the returned matrix is the
+    pairwise distance between the arrays from both `X` and `Y`.
+
+    Valid values for metric are:
+
+    - From scikit-learn: ['cityblock', 'cosine', 'euclidean', 'l1', 'l2', \
+        'manhattan'].
+    - From scipy.spatial.distance: ['sqeuclidean', 'canberra', 'minkowski', \
+        'jaccard', 'chebyshev']
+        See the documentation for scipy.spatial.distance for details on these
+        metrics.
+    - ['inner_product', 'hellinger']
+    Parameters
+    ----------
+    X : array-like (device or host) of shape (n_samples_x, n_features)
+        Acceptable formats: SciPy or Cupy sparse array
+
+    Y : array-like (device or host) of shape (n_samples_y, n_features),\
+        optional
+        Acceptable formats: SciPy or Cupy sparse array
+
+    metric : {"cityblock", "cosine", "euclidean", "l1", "l2", "manhattan", \
+        "sqeuclidean", "canberra", "lp", "inner_product", "minkowski", \
+        "jaccard", "hellinger", "chebyshev", "linf"}
+        The metric to use when calculating distance between instances in a
+        feature array.
+
+    convert_dtype : bool, optional (default = True)
+        When set to True, the method will, when necessary, convert
+        Y to be the same data type as X if they differ. This
+        will increase memory used for the method.
+
+    metric_arg : float, optional (default = 2)
+        Additionnal metric-specific argument.
+        For Minkowski it's the p-norm to apply.
+
+    Returns
+    -------
+    D : array [n_samples_x, n_samples_x] or [n_samples_x, n_samples_y]
+        A dense distance matrix D such that D_{i, j} is the distance between
+        the ith and jth vectors of the given matrix `X`, if `Y` is None.
+        If `Y` is not `None`, then D_{i, j} is the distance between the ith
+        array from `X` and the jth array from `Y`.
+
+    Examples
+    --------
+        >>> import cupy as cp
+        >>> from cuml.metrics import sparse_pairwise_distances
+        >>> import scipy
+        >>> from scipy import sparse
+        >>>
+        >>> X = scipy.sparse.random(2, 3, density=0.3)
```

Comment:
We should use cupy / device-based alternatives to scipy/numpy/sklearn in these examples wherever possible.